mailpoet/lib/Logging/PluginVersionProcessor.php

```diff
@@ -0,0 +1,13 @@
+<?php declare (strict_types=1);
+
+namespace MailPoet\Logging;
+
+use MailPoet\Config\Env;
+
+class PluginVersionProcessor {
```

Comment:
Any reason why this processor doesn't extend `ProcessorInterface` as the other processors?

Reply:
There are two reasons: 
- it is not mandatory to implement the interface, the method `pushProcessor` only requires a `callable`
- I tried that, but PHPStan complained about the `Record`:
```
  15     Parameter $record of method
         MailPoet\Logging\PluginVersionProcessor::__invoke() has invalid type
         MailPoetVendor\Monolog\Processor\Record.
  15     Return type (array) of method
         MailPoet\Logging\PluginVersionProcessor::__invoke() should be
         compatible with return type (MailPoetVendor\Monolog\Processor\Record)
         of method
         MailPoetVendor\Monolog\Processor\ProcessorInterface::__invoke()
```
I tried to fix it for some time but was unable. Rosta tried to help me but even he couldn't figure it out. So I removed the `ProcessorInterface` from this class.

Reply:
Thanks for the additional information, @pavel-mailpoet.

You probably saw that already, but I guess the errors are happening because of those two lines that use PHPStan specific documentation:

https://github.com/Seldaek/monolog/blob/2.4.0/src/Monolog/Processor/ProcessorInterface.php#L26-L27

On a quick look to me, it seems that the values of @phpstan-param and @phpstan-return are simply wrong, as the method signature clearly indicates that it receives an array and not Record. Have you considered implementing the interface and ignoring the PHPStan errors? I slightly prefer this approach as it would mean it would be easier to catch changes to the interface in the future (errors would happen on compile time instead of runtime). For example, this interface already changed in the latest version of Monolog:

https://github.com/Seldaek/monolog/blob/main/src/Monolog/Processor/ProcessorInterface.php#L26

What are your thoughts?

Reply:
> You probably saw that already, but I guess the errors are happening because of those two lines that use PHPStan specific documentation:

correct

> On a quick look to me, it seems that the values of @phpstan-param and @phpstan-return are simply wrong, as the method signature clearly indicates that it receives an array and not Record. Have you considered implementing the interface and ignoring the PHPStan errors? I slightly prefer this approach as it would mean it would be easier to catch changes to the interface in the future (errors would happen on compile time instead of runtime). For example, this interface already changed in the latest version of Monolog:

I am not a big fan of ignoring errors. It adds unnecessary stuff to the source code and can hide real errors. 

I am still not sure why should we implement the interface. The `pushProcessor` does not require it. It accepts anything `callable`. So I am not worried if the interface changes.

Reply:
Correct me if I'm wrong, but although the interface is not explicitly required when `pushProcessor()` is called, it seems to me it is implicitly required in other places. For example, when the processor is invoked the type of the parameter that is passed to it matters and without using the interface we won't know at compile time if the type changes and we might run into runtime errors. Considering this, I'm inclined to think that it is best to use the interface and ignore the specific PHPStan errors than to not use it. But it is a tradeoff and I understand your point about not being a fan of ignoring errors. In this case, I don't mind it so much as we are ignoring an error that is not our fault, it seems to be a problem in Monolog's code documentation.

Thoughts?

Reply:
Let me confirm this with you first. This is the change you suggest I make, right? https://d.pr/i/Jz7D6P – it makes the error go away, and PHPStan is green. 

I am not sure how it helps here, but if it helps could you please let me understand?

Reply:
>  This is the change you suggest I make, right?

This is close to what I had in mind, but instead of using `@phpstan-ignore-next-line`, we could add the error to `phpstan.neon`. This way, we are ignoring a specific error instead of all errors in the next line, and we get notified by PHPStan when the error is not present anymore or changes.

> I am not sure how it helps here, but if it helps could you please let me understand?

To me, the main advantage, as I tried to elaborate above, is that this allows us to extend `ProcessorInterface`, and thus we get errors at compile time instead of runtime if the interface changes. Happy to chat synchronously about this if you would like to.